testng-core-api/src/main/java/org/testng/internal/KeyAwareAutoCloseableLock.java

```diff
@@ -0,0 +1,37 @@
+package org.testng.internal;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A simple abstraction over {@link java.util.concurrent.locks.ReentrantLock} that can be used when
+ * we need to be dealing with a dictionary of lockable objects wherein we traditionally would have
+ * used the <code>synchronized</code> keyword.
+ */
+public final class KeyAwareAutoCloseableLock {
+  private final Map<Object, AutoCloseableLock> internalMap = new ConcurrentHashMap<>();
+
+  public AutoReleasable lockForObject(Object key) {
+    AutoCloseableLock internal =
+        internalMap.computeIfAbsent(Objects.requireNonNull(key), k -> new AutoCloseableLock());
+    return new AutoReleasable(internal.lock(), () -> internalMap.remove(key));
```

Comment:
This will likely fail with nested (reentrant) locking scenarios (e.g. `try(lock) { try(lock) }`). At least it is worth documenting.

Reply:
I didnt understand this part @vlsi . The method `lockForObject()` returns a `AutoReleasable` object whose constructor is package visible. Can you please help share a sample code wherein this can fail ? That will help me get a better understanding of the issue that you are calling out.

Reply:
Could you add a test for `KeyAwareAutoCloseableLock` that will test reentrant lock aqcuire?
The issue I mention is that if the usage is reentrant, then `internalMap` will be removed too early:

```java
Object key = ...;
KeyAwareAutoCloseableLock lock = ....;
try (... = lock.lockForObject(key)) {
    try (... = lock.lockForObject(key)) { // reentrant lock
    }
    // Here the entry will no longer be in the map
    // The next lockForObject will create a brand new AutoCloseableLock since the key is no longer in the map
    try (... = lock.lockForObject(key)) {
    }
}
```

Reply:
I think you are hinting at the discrepancy that I see in the output in terms of the `hashCode()` value being different for `ignore1` and `ignore2` but the end user of this class would expect them to be the same, which is NOT the case. Did you get it correctly @vlsi ?

Reply:
I would suggest something like `assertEquals(ignore1, ignore2)` and implement `KeyAwareAutoCloseableLock.AutoReleasable.equals` so it delegates to `ReentrantLock.equals`

Reply:
There's `java.util.concurrent.locks.ReentrantLock#isHeldByCurrentThread`.
In case the lock was already held when `lockForObject` executes, then you do not need to remove the key from the map (pass an empty lambda there)

Something like this might work:

```suggestion
    return new AutoReleasable(internal.lock(), internal.isHeldByCurrentThread() ? () -> {} : () -> internalMap.remove(key));
```

Reply:
Thank you for that suggestion. Trying to determine if the lock is already held by the current thread and skipping the removal part is basically causing this test to fail and thus break the build `test.github1461.MemoryLeakTestNg#testMemoryLeak` Trying to figure out what is going on.

Reply:
@vlsi - I have found why that test fails. That test has an expectation that within 10 seconds, all the test class instances will be gc'ed. In my case, since I am using the test class instance as my key, the object is still referenced in a static map and thus preventing the gc from happening. I have flipped my logic to start using the `hashCode()` obtained from the `toString()` representation of the test class (I chose this so that I dont end up with failures when `hashCode` is hard coded via JVM argument).

Reply:
I have pushed an additional commit that adds a test as per the suggestions made by you @vlsi ( That was a very helpful one. Thank you for that). Please check now.